// translation-check/dist/cjs/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty__default['default'](e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var url = 'https://mini.locize.com';
var showTranslations = function showTranslations(i18next) {
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  var isI18next = i18next.store && i18next.store.toJSON;
  var opts = {
    sourceLng: options.sourceLng,
    targetLngs: options.targetLngs,
    preserveEmptyStrings: options.preserveEmptyStrings,
    i18nFormat: options.i18nFormat
  };
  var resources;
  if (isI18next) {
    if (!opts.sourceLng) opts.sourceLng = i18next.languages[0];
    if (!opts.targetLngs) opts.targetLngs = i18next.options.supportedLngs ? i18next.options.supportedLngs.filter(function (l) {
      return l !== i18next.languages[0] && l !== 'cimode';
    }) : undefined;
    if (!opts.preserveEmptyStrings) opts.preserveEmptyStrings = i18next.options.returnEmptyString;
    if (!opts.i18nFormat) opts.i18nFormat = i18next.options.compatibilityJSON && i18next.options.compatibilityJSON !== 'v4' ? 'i18next_v3' : 'i18next_v4';
    resources = i18next.store.toJSON();
  } else {
    resources = i18next;
  }
  var newWindow = window.open(url);
  if (!newWindow) throw new Error('Please allow popups in your browser settings (https://support.google.com/chrome/answer/95472)');
  var initInt = setInterval(function () {
    return newWindow.postMessage({
      message: 'isEditorReady'
    }, url);
  }, 1000);
  window.addEventListener('message', function (evt) {
    if (evt.origin !== url) return;
    if (evt.data.message === 'editorIsReady') {
      clearInterval(initInt);
      newWindow.postMessage(_objectSpread(_objectSpread({
        message: 'initializeProject'
      }, opts), {}, {
        resources: resources
      }), url);
    }
  });
};
var defaultOptions = {
  queryStringParam: 'showtranslations'
};
var i18nextPlugin = {
  type: '3rdParty',
  init: function init(i18next) {
    var options = _objectSpread(_objectSpread({}, defaultOptions), i18next.options.translationStats);
    i18next.on('initialized', function () {
      var params = new URLSearchParams(window.location.search);
      if (params.has(options.queryStringParam)) {
        showTranslations(i18next, options);
      }
    });
  }
};

exports.i18nextPlugin = i18nextPlugin;
exports.showTranslations = showTranslations;
